### PR TITLE
Allow users to choose HTTP response buffer size

### PR DESCRIPTION
**Pull request by andrew-stevenson-sociomantic**
In order to minimise GC activity it is useful to pre-allocate buffers to a size that is a reasonable estimate of the largest size that will be seen. This change allows clients to do just that without changing the behaviour of existing clients.

### CODE REVIEW

**Review by mihails-strasuns-sociomantic — APPROVED** [review]
Looks good on its own though I wonder if others want to pursue this path as until now "reuse & grow automatically" seemed to be a most dominant pattern.

**Comment by andrew-stevenson-sociomantic** [comment]
> "reuse & grow automatically"

that still works and we still use it but if you know there will be a grow within the first seconds...

**Comment by gavin-norman-sociomantic** [comment]
One question about wording: you use the word "clients", but we don't have an HTTP client in ocean. What are you referring to exactly (this seems to all be HTTP server-side code)?

**Comment by nemanja-boric-sociomantic** [comment]
I assumed the users of this facility.

**Comment by gavin-norman-sociomantic** [comment]
>I assumed the users of this facility.

I guess so too, but using the word "client" in the context of HTTP is kind of loaded ;)

**Comment by andrew-stevenson-sociomantic** [comment]
Yes Nemanja is correct, a client is a user of provided feature. Is it not the ordinary English meaning of client?

**Comment by gavin-norman-sociomantic** [comment]
It's not the usual library terminology. We usually refer to "users" of library code. (Plus the confusion over HTTP clients.)

**Review by gavin-norman-sociomantic — COMMENTED** [review]
Change "clients" to "users".

**Comment by andrew-stevenson-sociomantic** [comment]
I have made the change - though I would have thought any possible ambiguity with client must also exist with user.

**Comment by nemanja-boric-sociomantic** [comment]
Oh, the last thing. This should be based on `v2.x.x` and not on `v3.x.x`. You can change the target branch above, but the rebase could also be due.

**Review by nemanja-boric-sociomantic — APPROVED** [review]

**Comment by gavin-norman-sociomantic** [comment]
>I have made the change - though I would have thought any possible ambiguity with client must also exist with user.

Who's ever heard of an "HTTP user"? ;)

File: src/ocean/net/Uri.d [hunk 1]
```diff
@@ -227,10 +227,10 @@ class Uri : UriView
 
         ***********************************************************************/
 
-        this ()
+        this ( uint initial_buffer_size = 512 )
```
Comment (gavin-norman-sociomantic) [inline src/ocean/net/Uri.d]: Method params doc should be updated.

File: src/ocean/net/http/HttpRequest.d [hunk 2]
```diff
@@ -170,14 +170,15 @@ class HttpRequest : HttpHeader
 
      **************************************************************************/
 
-    public this ( bool add_entity_headers = false, size_t msg_body_prealloc_length = 0 )
+    public this ( bool add_entity_headers = false, size_t msg_body_prealloc_length = 0,
```
Comment (gavin-norman-sociomantic) [inline src/ocean/net/http/HttpRequest.d:177]: Method params doc should be updated.

**Review by gavin-norman-sociomantic — COMMENTED** [review]
Tiny doc comment.

**Comment by andrew-stevenson-sociomantic** [comment]
Updated comments as per Gavin's request. Fixed some of the existing wording whilst I was there.

**Review by gavin-norman-sociomantic — APPROVED** [review]
Thanks!

**Pull request merged**